tools/buildsys/src/gomod.rs

```diff
@@ -0,0 +1,160 @@
+
+use super::manifest;
+use duct::cmd;
+use snafu::{OptionExt, ResultExt};
+use std::env;
+use std::path::PathBuf;
+use std::process::Output;
+
+pub(crate) struct GoMod;
+
+impl GoMod {
+    pub(crate) fn vendor(
+        root_dir: &PathBuf,
+        package_dir: &PathBuf,
+        gomods: &[manifest::GoModule],
+    ) -> Result<Self> {
+        for g in gomods {
+            let input_path_arg = g.input.as_ref().context(error::InputFileSnafu)?;
+            let input_path = package_dir.join(input_path_arg);
+            if !input_path.is_file() {
+                return Err(error::Error::InputFileBad);
+            }
+            let mod_dir = g.mod_dir.as_ref().context(error::ModDirSnafu)?;
+            let output_dir_arg = g.output_dir.as_ref().context(error::OutputDirSnafu)?;
+            let output_dir = package_dir.join(output_dir_arg);
+            if output_dir.exists() && !output_dir.is_dir() {
+                return Err(error::Error::OutputDirBad);
+            }
+
+            // Our SDK and toolchain are picked by the external `cargo make` invocation.
+            let sdk = getenv("BUILDSYS_SDK_IMAGE")?;
+
+            // Several Go variables control proxying
+            let goproxy = go_env("GOPROXY").unwrap_or("".to_string());
+            let gosumdb = go_env("GOSUMDB").unwrap_or("".to_string());
+            let goprivate = go_env("GOPRIVATE").unwrap_or("".to_string());
+
+            let args = DockerGoArgs {
+                module_path: package_dir.clone(),
+                sdk_image: sdk,
+                go_mod_cache: root_dir.join(".gomodcache".to_string()),
+                command: format!(
+                    "mkdir -p {outdir}
+                    tar zxf {input} -C {outdir} &&
+                    cd {outdir}/{moddir} &&
+                    export GOPROXY={goproxy} &&
+                    export GOSUMDB={gosumdb} &&
+                    export GOPRIVATE={goprivate} &&
+                    go list -mod=readonly ./... >/dev/null && go mod vendor",
+                    input = input_path_arg.to_string_lossy(),
+                    outdir = output_dir_arg.to_string_lossy(),
+                    moddir = mod_dir.to_string_lossy(),
+                    goproxy = goproxy,
+                    gosumdb = gosumdb,
+                    goprivate = goprivate,
+                ),
+            };
+            match docker_go(root_dir, &args) {
+                Ok(_) => continue,
+                Err(e) => return Err(e),
+            }
+        }
+
+        return Ok(Self);
+    }
+}
+
+struct DockerGoArgs {
+    module_path: PathBuf,
+    sdk_image: String,
+    go_mod_cache: PathBuf,
+    command: String,
+}
+
+/// Run `docker-go` with the specified arguments.
+fn docker_go(root_dir: &PathBuf, dg_args: &DockerGoArgs) -> Result<Output> {
+    let args = vec![
+        "--module-path",
+        dg_args
+            .module_path
+            .to_str()
+            .context(error::InputFileSnafu)
+            .unwrap(),
```

Comment:
Hmm why do we still need to `unwrap()` here?

Generally that would mean that we would add another `.context(error::WhateverSnafu)` to the chain.